pykern/fconf.py

```diff
@@ -0,0 +1,517 @@
+just use self to call something::
+
+    def count_up(self, top):
+        return reversed(self.count_down(top))
+
+Note that the above count_down example used a special feature of
+result merging. When a Macro returns a list (generator, tuple, etc.)
+in a list context, it will merge with the list. Strings are used in
+place. For example::
+
+    mixed:
+      - range(3)
+      - make_uri('sirepo.com')
+
+Produces::
+
+    {'mixed': [0, 1, 2, 'https://sirepo.com']}
+
+The same is true of dicts. For example::
+
+    fconf_macros:
+      base(num):
+        v${num}.radia.run:
+          description: VM ${num}
+
+    all_hosts:
+      base(3):
+      base(5):
+
+Produces::
+
+    {
+        "all_hosts": {
+            "v3.radia.run": {
+                "description": "VM 3",
+            },
+            "v5.radia.run": {
+                "description": "VM 5",
+            },
+        },
+    }
+
+There are some other useful features. You can refer to any previous
+declared value using variable expansion. For example::
+
+    a:
+      b:
+        c: 3
+    d: ${a}
+    e: ${a.b.c}
+
+
+Produces::
+
+    {
+        'a': {'b': {'c': 3}},
+        'd': {'b': {'c': 3}},
+        'e': 3,
+    }
+
+This allows for flexible constants. Those global strings can be used
+in YAML Macros.
+
+YAML files are evaluated before they are merged. However, they are all
+merged before the next file is evaluated. This allows a main
+"constants" file, for example, to direct the flow of the subsequent files.
+
+:copyright: Copyright (c) 2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+
+"""
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdlog, pkdp
+import contextlib
+import copy
+import inspect
+import pykern.pkrunpy
+import re
+import pykern.pkyaml
+import pykern.pkcollections
+import pykern.pkio
+
+
+#: parse_files macro expansion pattern
+_TEXT_MACRO_NAME = re.compile(
+    r"^([a-z]\w*)\(((?:\s*[a-z]\w*\s*)?(?:,\s*[a-z]\w*\s*)*)\)$", flags=re.IGNORECASE
+)
+
+_MACRO_CALL = re.compile(r"^([a-z]\w*)\((.*)\)$", flags=re.IGNORECASE + re.DOTALL)
+
+_ARG_SEP = re.compile(r"\s*,\s*|\s+")
+
+_FVAR = re.compile(r"\$\{(\S+)\}")
+_FVAR_EXACT = re.compile(f"^{_FVAR.pattern}$")
+
+_SELF = "fconf_self"
+
+_NO_PARAM = object()
+
+
+def parse_all(path):
+    """Parse all the Python and YAML files in `directory`
+
+    Args:
+        path (py.path): directory that *.py and *.yml files
+    """
+    return Parser(
+        pykern.pkio.sorted_glob(path.join("*.py"))
+        + pykern.pkio.sorted_glob(path.join("*.yml")),
+    ).result
+
+
+class Parser(PKDict):
+    def __init__(self, files):
+        self.pkupdate(
+            files=PKDict(py=[], yml=[]),
+            macros=PKDict(),
+        )
+        for f in files:
+            try:
+                self._add_file(f)
+            except Exception:
+                pkdlog("error parsing file={}", f)
```

Comment:
What about using [`raise from`](https://stackoverflow.com/a/46091127/5518313) ? I could see myself not seeing the pkdlog above the error.

Reply:
Thanks! I didn't know about raise from. 6a9b1c0